source/extensions/filters/http/buffer/config.cc

```diff
@@ -17,7 +17,18 @@ namespace BufferFilter {
 
 absl::StatusOr<Http::FilterFactoryCb> BufferFilterFactory::createFilterFactoryFromProtoTyped(
     const envoy::extensions::filters::http::buffer::v3::Buffer& proto_config, const std::string&,
-    DualInfo, Server::Configuration::ServerFactoryContext&) {
+    DualInfo, Server::Configuration::FactoryContext&) {
+  return createFilterFactory(proto_config);
+}
+
+absl::StatusOr<Http::FilterFactoryCb> BufferFilterFactory::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::filters::http::buffer::v3::Buffer& proto_config, const std::string&,
+    DualInfo, Server::Configuration::UpstreamFactoryContext&) {
+  return createFilterFactory(proto_config);
+}
+
+absl::StatusOr<Http::FilterFactoryCb> BufferFilterFactory::createFilterFactory(
+    const envoy::extensions::filters::http::buffer::v3::Buffer& proto_config) {
```

Comment:
To convert the match_delegate filter into dual filter, I have to change factory_base.h to have two interface of createFilterFactoryFromProtoTyped(), one for Server::Configuration::ServerFactoryContext, and one for Server::Configuration::UpstreamFactoryContext&.  Thus I have to change all the existing dual filter implementation.

Reply:
I see. I think the change is because the `match_delegate` needs to access the original FactoryContext and UpstreamFactoryContext.

Then, rather than changing the DualFactoryBase's methods here, I prefer just to change the `match_delegate` to avoid inheriting from `DualFactoryBase`.
Thats say, the `match_delegate` could inherit the `NamedHttpFilterConfigFactory ` and `UpstreamHttpFilterConfigFactory `.

If we do think the `DualFactoryBase` should update it's methods, we can do the refactoring in separated PR.

Reply:
My 2 cents: I agree with @wbpcode that separate inheritance from `NamedHttpFilterConfigFactory`  and `UpstreamHttpFilterConfigFactory` is probably a better idea (if this approach works).

Given the uniqueness of match_delegate filter, i think other filters won't need this new API. we can make match_delegate as an exception to that doc.

Reply:
I don't have strong preference on this. However, if we go that route, and if another filter also has logic based on factory type(NamedHttpFilterConfigFactory/UpstreamHttpFilterConfigFactory), then we will have to make it another exception if we want to convert into dual filter.

Overall, I feel the way current DualFactoryBase method: createFilterFactoryFromProtoTyped() implementation, which only support parameter ServerFactoryContext,  greatly limited its flexibility.  We should let it be able to support both FactoryContext and UpstreamFactoryContext, like createFilterFactoryFromProto().

Reply:
Kind ping on this!

Reply:
Yeah. The `match_delegate` could be a special case. The docs just a recommend way to develop a new dual filter and not a strict rule.

Reply:
cc @yanjunxiang-google

Reply:
I created a new PR for this: https://github.com/envoyproxy/envoy/pull/33013.  In the new PR, the match_delegate filter is inherited from NamedHttpFilterConfigFactory and UpstreamHttpFilterConfigFactory , kind like DualFactoryBase.